### PR TITLE
Avoid deadlock between Put and Get

### PR DESCRIPTION
**Pull request by stevenosborne-wf**
There is a problem where the batch channel fills up and Puts hold the lock preventing Gets from catching up.

@dustinhiatt-wf @brianshannan-wf @tylertreat @alexandercampbell-wf

### CODE REVIEW

**Comment by alexandercampbell-wf** [comment]
+1

File: batcher/batcher.go [hunk 1]
```diff
@@ -173,7 +187,13 @@ func (b *basicBatcher) IsDisposed() bool {
...
 // flush is not threadsafe, so should be synchronized externally.
 func (b *basicBatcher) flush() {
-	b.batchChan <- b.items
+	// Note: This needs to be in a go-routine to avoid locking out gets when
+	// the batch channel is full.
+	cpItems := make([]interface{}, len(b.items))
+	for i, val := range b.items {
+		cpItems[i] = val
+	}
+	go func() { b.batchChan <- cpItems }()
```
Comment (brianshannan-wf) [inline batcher/batcher.go]: This wouldn't necessarily preserve the order of items coming in versus going out, as go routines could add the batches in any order once they get backed up, or really before too.

**Comment by brianshannan-wf** [comment]
+1 if we don't care about ordering per above comment

File: batcher/batcher.go [hunk 2]
```diff
@@ -158,6 +161,17 @@ func (b *basicBatcher) Dispose() {
 	b.flush()
 	b.disposed = true
 	b.items = nil
+
+	// Drain the batch channel and all routines waiting to put on the channel
+DrainLoop:
+	for {
+		select {
+		case <-b.batchChan:
+		case <-time.After(5 * time.Millisecond):
```
Comment (tylertreat-wf) [inline batcher/batcher.go]: Seems like there could be a more intelligent/safer way to know if anyone is waiting to put stuff on the channel.

File: batcher/batcher.go (hunk 1 above)
Comment (tylertreat-wf) [inline batcher/batcher.go]: I don't think we care about the ordering, but might want to document it.

**Comment by stevenosborne-wf** [comment]
@brianshannan-wf @tylertreat-wf Comments addressed. @alexandercampbell-wf Can you review the most recent commits?

**Comment by alexandercampbell-wf** [comment]
LGTM but I don't know much about this code.

File: batcher/batcher.go [hunk 3]
```diff
@@ -158,6 +165,11 @@ func (b *basicBatcher) Dispose() {
 	b.flush()
 	b.disposed = true
 	b.items = nil
+
+	// Drain the batch channel and all routines waiting to put on the channel
+	for len(b.batchChan) > 0 || atomic.LoadInt32(&b.waiting) > 0 {
```
Comment (dustinhiatt-wf) [inline batcher/batcher.go]: Not sure if you'll ever hit it, or if it matters, but two threads could have len(b.batchChan) evaluate to greater than 0 at the same time and if there was only one the other will be blocked on the next line.

Comment (stevenosborne-wf) [inline batcher/batcher.go]: We're safe in here because of the lock.

File: batcher/batcher.go [hunk 4]
```diff
@@ -158,6 +165,11 @@ func (b *basicBatcher) Dispose() {
...
+		<-b.batchChan
+	}
 	close(b.batchChan)
```
Comment (dustinhiatt-wf) [inline batcher/batcher.go]: So what happens here if 1000 goroutines are waiting to send on this channel?  Does only one get cleared before the rest panic on this close?

Comment (stevenosborne-wf) [inline batcher/batcher.go]: Each routine bumps ```b.waiting``` to ensure the drain runs enough times. See [here](https://github.com/Workiva/go-datastructures/blob/batcher_deadlock/batcher/batcher.go#L195).

**Comment by dustinhiatt-wf** [comment]
+1

**Comment by brianshannan-wf** [comment]
+1

**Comment by stevenosborne-wf** [comment]
Investigating regression.

**Comment by tylertreat-wf** [comment]
+1

**Comment by stevenosborne-wf** [comment]
@dustinhiatt-wf @brianshannan-wf @alexandercampbell-wf Fix deadlock in dispose. Ready for review.

**Comment by dustinhiatt-wf** [comment]
+1

**Comment by brianshannan-wf** [comment]
+1

**Pull request merged**